File: js/BFFFlattener.js
/**
 * BFF展开器 - JavaScript接口
 * 
 * 支持两种模式：
 * 1. WASM模式：使用C++编译的WebAssembly，性能更高
 * 2. 纯JS模式：WASM不可用时的备选方案
 */

export class BFFFlattener {
    constructor() {
        this.wasmModule = null;
        this.isWasmReady = false;
        this.useWasm = false;
        
        // 纯JS备选数据
        this.vertices = [];
        this.faces = [];
        this.seamEdges = new Set();
        this.uvResult = null;
    }
    
    /**
     * 初始化（尝试加载WASM模块）
     */
    async init() {
        try {
            // 尝试动态导入WASM模块
            if (typeof BFFModule !== 'undefined') {
                this.wasmModule = await BFFModule();
                this.wasmModule.init();
                this.isWasmReady = true;
                this.useWasm = true;
                console.log('BFF: 使用WASM加速模式');
                return true;
            }
        } catch (e) {
            console.warn('BFF: WASM模块加载失败，使用纯JavaScript模式', e);
        }
        
        console.log('BFF: 使用纯JavaScript模式');
        this.useWasm = false;
        return true;
    }
    
    /**
     * 设置网格数据
     * @param {Array} vertices - 顶点数组 [{x, y, z}, ...]
     * @param {Array} faces - 面数组 [[v0, v1, v2], ...]
     */
    setMesh(vertices, faces) {
        if (this.useWasm && this.wasmModule) {
            // 转换为扁平数组
            const flatVerts = new Float64Array(vertices.length * 3);
            for (let i = 0; i < vertices.length; i++) {
                flatVerts[i * 3] = vertices[i].x;
                flatVerts[i * 3 + 1] = vertices[i].y;
                flatVerts[i * 3 + 2] = vertices[i].z;
            }
            
            const flatFaces = new Int32Array(faces.length * 3);
            for (let i = 0; i < faces.length; i++) {
                flatFaces[i * 3] = faces[i][0];
                flatFaces[i * 3 + 1] = faces[i][1];
                flatFaces[i * 3 + 2] = faces[i][2];
            }
            
            this.wasmModule.setMesh(flatVerts, flatFaces);
        } else {
            this.vertices = vertices;
            this.faces = faces;
        }
    }
    
    /**
     * 添加缝线边
     * @param {number} v1 - 顶点1索引
     * @param {number} v2 - 顶点2索引
     */
    addSeamEdge(v1, v2) {
        if (this.useWasm && this.wasmModule) {
            this.wasmModule.addSeamEdge(v1, v2);
        } else {
            const key = v1 < v2 ? `${v1}_${v2}` : `${v2}_${v1}`;
            this.seamEdges.add(key);
        }
    }
    
    /**
     * 清除缝线
     */
    clearSeams() {
        if (this.useWasm && this.wasmModule) {
            this.wasmModule.clearSeams();
        } else {
            this.seamEdges.clear();
        }
    }
    
    /**
     * 执行展开
     * @param {Function} onProgress - 进度回调
     * @returns {Promise<Object>} 展开结果
     */
    async flatten(onProgress = null) {
        if (this.useWasm && this.wasmModule) {
            return this.flattenWasm(onProgress);
        } else {
            return this.flattenJS(onProgress);
        }
    }
    
    /**
     * WASM展开
     */
    async flattenWasm(onProgress) {
        if (onProgress) onProgress(10);
        
        const success = this.wasmModule.flatten();
        
        if (!success) {
            throw new Error(this.wasmModule.getError());
        }
        
        if (onProgress) onProgress(90);
        
        const uvArray = this.wasmModule.getUVCoords();
        const uvCount = this.wasmModule.getUVCount();
        
        // 转换为UV对象数组
        const uvs = [];
        for (let i = 0; i < uvCount; i++) {
            uvs.push({
                u: uvArray[i * 2],
                v: uvArray[i * 2 + 1]
            });
        }
        
        if (onProgress) onProgress(100);
        
        return {
            uvs: uvs,
            success: true
        };
    }
    
    /**
     * 纯JavaScript展开（优化版）
     */
    async flattenJS(onProgress) {
        if (this.vertices.length === 0 || this.faces.length === 0) {
            throw new Error('Empty mesh');
        }
        
        const uvs = new Array(this.vertices.length).fill(null).map(() => ({ u: 0, v: 0 }));
        const placed = new Set();
        
        if (onProgress) onProgress(5);
        
        // 预计算边到面的映射
        const edgeToFaces = new Map();
        for (let fIdx = 0; fIdx < this.faces.length; fIdx++) {
            const face = this.faces[fIdx];
            for (let i = 0; i < 3; i++) {
                const v1 = face[i];
                const v2 = face[(i + 1) % 3];
                const key = v1 < v2 ? `${v1}_${v2}` : `${v2}_${v1}`;
                if (!edgeToFaces.has(key)) {
                    edgeToFaces.set(key, []);
                }
                edgeToFaces.get(key).push(fIdx);
            }
        }
        
        if (onProgress) onProgress(10);
        
        // 放置第一个三角形
        const firstFace = this.faces[0];
        const v0 = this.vertices[firstFace[0]];
        const v1 = this.vertices[firstFace[1]];
        const v2 = this.vertices[firstFace[2]];
        
        const e01 = this.distance(v0, v1);
        const e02 = this.distance(v0, v2);
        const e12 = this.distance(v1, v2);
        
        uvs[firstFace[0]] = { u: 0, v: 0 };
        placed.add(firstFace[0]);
        
        uvs[firstFace[1]] = { u: e01, v: 0 };
        placed.add(firstFace[1]);
        
        const cosA = Math.max(-1, Math.min(1, (e01 * e01 + e02 * e02 - e12 * e12) / (2 * e01 * e02)));
        const sinA = Math.sqrt(1 - cosA * cosA);
        uvs[firstFace[2]] = { u: e02 * cosA, v: e02 * sinA };
        placed.add(firstFace[2]);
        
        // BFS展开
        const processedFaces = new Set([0]);
        const queue = [0];
        let processCount = 0;
        const totalFaces = this.faces.length;
        
        while (queue.length > 0) {
            const currentFaceIdx = queue.shift();
            const currentFace = this.faces[currentFaceIdx];
            
            processCount++;
            if (onProgress && processCount % 100 === 0) {
                onProgress(10 + (processCount / totalFaces) * 70);
                await new Promise(r => setTimeout(r, 0)); // 让UI响应
            }
            
            // 检查每条边的相邻面
            for (let i = 0; i < 3; i++) {
                const ev1 = currentFace[i];
                const ev2 = currentFace[(i + 1) % 3];
                const key = ev1 < ev2 ? `${ev1}_${ev2}` : `${ev2}_${ev1}`;
                
                const neighborFaces = edgeToFaces.get(key) || [];
                
                for (const neighborFaceIdx of neighborFaces) {
                    if (processedFaces.has(neighborFaceIdx)) continue;
                    
                    const neighborFace = this.faces[neighborFaceIdx];
                    
                    // 找共享顶点和新顶点
                    let sharedV1 = -1, sharedV2 = -1, newV = -1;
                    for (const v of neighborFace) {
                        if (placed.has(v)) {
                            if (sharedV1 === -1) sharedV1 = v;
                            else sharedV2 = v;
                        } else {
                            newV = v;
                        }
                    }
                    
                    if (sharedV1 === -1 || sharedV2 === -1 || newV === -1) continue;
                    
                    // 计算新顶点UV
                    const p1 = uvs[sharedV1];
                    const p2 = uvs[sharedV2];
                    
                    const len12 = this.distance(this.vertices[sharedV1], this.vertices[sharedV2]);
                    const len1n = this.distance(this.vertices[sharedV1], this.vertices[newV]);
                    const len2n = this.distance(this.vertices[sharedV2], this.vertices[newV]);
                    
                    if (len12 < 1e-10) continue;
                    
                    const cosAngle = Math.max(-1, Math.min(1, 
                        (len12 * len12 + len1n * len1n - len2n * len2n) / (2 * len12 * len1n)));
                    const sinAngle = Math.sqrt(1 - cosAngle * cosAngle);
                    
                    // 计算方向
                    const dx = p2.u - p1.u;
                    const dy = p2.v - p1.v;
                    const len2d = Math.sqrt(dx * dx + dy * dy);
                    
                    if (len2d < 1e-10) continue;
                    
                    const dirX = dx / len2d;
                    const dirY = dy / len2d;
                    const perpX = -dirY;
                    const perpY = dirX;
                    
                    // 新位置 (newVertexIdx 是顶点索引，newUVu/newUVv 是UV坐标)
                    const newVertexIdx = newV; // 保存顶点索引
                    let newUVu = p1.u + dirX * len1n * cosAngle + perpX * len1n * sinAngle;
                    let newUVv = p1.v + dirY * len1n * cosAngle + perpY * len1n * sinAngle;
                    
                    // 检查朝向
                    const cross = (p2.u - p1.u) * (newUVv - p1.v) - (p2.v - p1.v) * (newUVu - p1.u);
                    if (cross < 0) {
                        newUVu = p1.u + dirX * len1n * cosAngle - perpX * len1n * sinAngle;
                        newUVv = p1.v + dirY * len1n * cosAngle - perpY * len1n * sinAngle;
                    }
                    
                    uvs[newVertexIdx] = { u: newUVu, v: newUVv };
                    placed.add(newVertexIdx);
                    processedFaces.add(neighborFaceIdx);
                    queue.push(neighborFaceIdx);
                }
            }
        }
        
        if (onProgress) onProgress(70);
        
        // 简单的平滑优化
        await this.optimizeUVs(uvs, 15, onProgress);
        
        if (onProgress) onProgress(85);
        
        // 检测UV岛并分离排列
        const islands = this.detectUVIslands(uvs);
        console.log(`检测到 ${islands.length} 个UV岛`);
        
        if (onProgress) onProgress(90);
        
        // 排列UV岛，确保不重叠
        this.arrangeUVIslands(uvs, islands);
        
        if (onProgress) onProgress(95);
        
        // 最终归一化
        this.normalizeUVs(uvs);
        
        if (onProgress) onProgress(100);
        
        this.uvResult = uvs;
        
        return {
            uvs: uvs,
            islands: islands,
            success: true
        };
    }
    
    /**
     * 检测UV岛 - 找出不连通的UV区域
     * @returns {Array} 每个UV岛包含的面索引数组
     */
    detectUVIslands(uvs) {
        const islands = [];
        const visitedFaces = new Set();
        
        // 构建面到面的邻接关系（通过共享边）
        const faceAdjacency = new Map();
        const edgeToFaces = new Map();
        
        for (let faceIdx = 0; faceIdx < this.faces.length; faceIdx++) {
            const face = this.faces[faceIdx];
            faceAdjacency.set(faceIdx, new Set());
            
            // 记录每条边属于哪些面
            for (let i = 0; i < 3; i++) {
                const v1 = face[i];
                const v2 = face[(i + 1) % 3];
                const edgeKey = v1 < v2 ? `${v1}_${v2}` : `${v2}_${v1}`;
                
                // 检查这条边是否是缝线（被切断）
                const isSeamEdge = this.seamEdges.has(edgeKey);
                
                if (!isSeamEdge) {
                    if (!edgeToFaces.has(edgeKey)) {
                        edgeToFaces.set(edgeKey, []);
                    }
                    edgeToFaces.get(edgeKey).push(faceIdx);
                }
            }
        }
        
        // 建立面邻接关系（排除缝线边）
        for (const [edgeKey, faces] of edgeToFaces) {
            if (faces.length === 2) {
                faceAdjacency.get(faces[0]).add(faces[1]);
                faceAdjacency.get(faces[1]).add(faces[0]);
            }
        }
        
        // BFS找出所有连通的UV岛
        for (let faceIdx = 0; faceIdx < this.faces.length; faceIdx++) {
            if (visitedFaces.has(faceIdx)) continue;
            
            const island = {
                faces: [],
                vertices: new Set()
            };
            
            const queue = [faceIdx];
            visitedFaces.add(faceIdx);
            
            while (queue.length > 0) {
                const currentFace = queue.shift();
                island.faces.push(currentFace);
                
                // 添加该面的顶点
                for (const v of this.faces[currentFace]) {
                    island.vertices.add(v);
                }
                
                // 遍历相邻面
                const neighbors = faceAdjacency.get(currentFace);
                if (neighbors) {
                    for (const neighbor of neighbors) {
                        if (!visitedFaces.has(neighbor)) {
                            visitedFaces.add(neighbor);
                            queue.push(neighbor);
                        }
                    }
                }
            }
            
            islands.push(island);
        }
        
        return islands;
    }
    
    /**
     * 排列UV岛，确保它们不重叠
     */
    arrangeUVIslands(uvs, islands) {
        if (islands.length <= 1) return;
        
        // 计算每个岛的边界框
        const islandBounds = islands.map(island => {
            let minU = Infinity, maxU = -Infinity;
            let minV = Infinity, maxV = -Infinity;
            
            for (const v of island.vertices) {
                if (uvs[v]) {
                    minU = Math.min(minU, uvs[v].u);
                    maxU = Math.max(maxU, uvs[v].u);
                    minV = Math.min(minV, uvs[v].v);
                    maxV = Math.max(maxV, uvs[v].v);
                }
            }
            
            return {
                minU, maxU, minV, maxV,
                width: maxU - minU,
                height: maxV - minV,
                area: (maxU - minU) * (maxV - minV)
            };
        });
        
        // 按面积从大到小排序岛
        const sortedIndices = islands.map((_, i) => i)
            .sort((a, b) => islandBounds[b].area - islandBounds[a].area);
        
        // 使用简单的行排列算法
        const padding = 0.05; // 岛之间的间距
        let currentX = 0;
        let currentY = 0;
        let rowHeight = 0;
        const maxRowWidth = 2.0; // 最大行宽
        
        for (const islandIdx of sortedIndices) {
            const island = islands[islandIdx];
            const bounds = islandBounds[islandIdx];
            
            // 检查是否需要换行
            if (currentX + bounds.width > maxRowWidth && currentX > 0) {
                currentX = 0;
                currentY += rowHeight + padding;
                rowHeight = 0;
            }
            
            // 计算偏移量，将岛移动到新位置
            const offsetU = currentX - bounds.minU;
            const offsetV = currentY - bounds.minV;
            
            // 应用偏移
            for (const v of island.vertices) {
                if (uvs[v]) {
                    uvs[v].u += offsetU;
                    uvs[v].v += offsetV;
                }
            }
            
            // 更新位置
            currentX += bounds.width + padding;
            rowHeight = Math.max(rowHeight, bounds.height);
        }
    }
    
    /**
     * UV优化（拉普拉斯平滑）
     */
    async optimizeUVs(uvs, iterations, onProgress) {
        // 构建顶点邻接
        const neighbors = new Map();
        for (const face of this.faces) {
            for (let i = 0; i < 3; i++) {
                const v = face[i];
                if (!neighbors.has(v)) neighbors.set(v, new Set());
                neighbors.get(v).add(face[(i + 1) % 3]);
                neighbors.get(v).add(face[(i + 2) % 3]);
            }
        }
        
        const alpha = 0.5;
        
        for (let iter = 0; iter < iterations; iter++) {
            const newUVs = uvs.map(uv => ({ ...uv }));
            
            for (const [v, nbs] of neighbors) {
                if (nbs.size < 2) continue;
                
                let sumU = 0, sumV = 0;
                for (const n of nbs) {
                    sumU += uvs[n].u;
                    sumV += uvs[n].v;
                }
                
                newUVs[v].u = uvs[v].u * (1 - alpha) + (sumU / nbs.size) * alpha;
                newUVs[v].v = uvs[v].v * (1 - alpha) + (sumV / nbs.size) * alpha;
            }
            
            for (let i = 0; i < uvs.length; i++) {
                uvs[i] = newUVs[i];
            }
            
            if (onProgress && iter % 5 === 0) {
                onProgress(85 + (iter / iterations) * 10);
                await new Promise(r => setTimeout(r, 0));
            }
        }
    }
    
    /**
     * 归一化UV坐标到 [0, 1]
     */
    normalizeUVs(uvs) {
        let minU = Infinity, maxU = -Infinity;
        let minV = Infinity, maxV = -Infinity;
        
        for (const uv of uvs) {
            if (uv.u < minU) minU = uv.u;
            if (uv.u > maxU) maxU = uv.u;
            if (uv.v < minV) minV = uv.v;
            if (uv.v > maxV) maxV = uv.v;
        }
        
        const scale = Math.max(maxU - minU, maxV - minV);
        if (scale < 1e-10) return;
        
        for (const uv of uvs) {
            uv.u = (uv.u - minU) / scale;
            uv.v = (uv.v - minV) / scale;
        }
    }
    
    /**
     * 计算3D距离
     */
    distance(v1, v2) {
        const dx = v2.x - v1.x;
        const dy = v2.y - v1.y;
        const dz = v2.z - v1.z;
        return Math.sqrt(dx * dx + dy * dy + dz * dz);
    }
    
    /**
     * 清理资源
     */
    dispose() {
        if (this.wasmModule) {
            this.wasmModule.cleanup();
            this.wasmModule = null;
        }
        this.vertices = [];
        this.faces = [];
        this.seamEdges.clear();
        this.uvResult = null;
    }
}

File: wasm/src/bff_flattener.cpp
/**
 * BFF (Boundary First Flattening) UV展开算法实现
 */

#include "bff_flattener.h"
#include <queue>
#include <cstring>
#include <limits>

namespace bff {

BFFFlattener::BFFFlattener() {
}

BFFFlattener::~BFFFlattener() {
}

void BFFFlattener::setMesh(const double* vertices, int numVertices,
                           const int* faces, int numFaces) {
    mesh.vertices.clear();
    mesh.faces.clear();
    mesh.halfEdges.clear();
    mesh.vertexHalfEdge.clear();
    mesh.isBoundaryVertex.clear();
    mesh.seamEdges.clear();
    uvResult.clear();
    
    // 复制顶点
    mesh.vertices.resize(numVertices);
    for (int i = 0; i < numVertices; i++) {
        mesh.vertices[i] = Vec3(vertices[i * 3], vertices[i * 3 + 1], vertices[i * 3 + 2]);
    }
    
    // 复制面（三角形）
    mesh.faces.resize(numFaces);
    for (int i = 0; i < numFaces; i++) {
        mesh.faces[i] = {faces[i * 3], faces[i * 3 + 1], faces[i * 3 + 2]};
    }
    
    mesh.vertexHalfEdge.resize(numVertices, -1);
    mesh.isBoundaryVertex.resize(numVertices, false);
    
    // 构建半边结构
    buildHalfEdgeStructure();
    identifyBoundaries();
}

void BFFFlattener::addSeamEdge(int v1, int v2) {
    auto key = edgeKey(v1, v2);
    if (mesh.seamEdges.empty()) {
        mesh.seamEdges.resize(1);
    }
    mesh.seamEdges[0].insert(key.first * 1000000 + key.second);
}

void BFFFlattener::clearSeams() {
    mesh.seamEdges.clear();
}

void BFFFlattener::buildHalfEdgeStructure() {
    mesh.halfEdges.clear();
    
    std::map<std::pair<int, int>, int> edgeToHalfEdge;
    
    for (int faceIdx = 0; faceIdx < (int)mesh.faces.size(); faceIdx++) {
        const auto& face = mesh.faces[faceIdx];
        int numVerts = face.size();
        int firstHE = mesh.halfEdges.size();
        
        for (int i = 0; i < numVerts; i++) {
            HalfEdge he;
            he.vertex = face[(i + 1) % numVerts];
            he.face = faceIdx;
            he.next = firstHE + (i + 1) % numVerts;
            he.prev = firstHE + (i + numVerts - 1) % numVerts;
            he.twin = -1;
            he.isBoundary = false;
            he.isSeam = false;
            
            int heIdx = mesh.halfEdges.size();
            mesh.halfEdges.push_back(he);
            
            // 记录边到半边的映射
            int v1 = face[i];
            int v2 = face[(i + 1) % numVerts];
            edgeToHalfEdge[edgeKey(v1, v2)] = heIdx;
            
            // 设置顶点的半边引用
            if (mesh.vertexHalfEdge[v1] == -1) {
                mesh.vertexHalfEdge[v1] = heIdx;
            }
        }
    }
    
    // 设置twin半边
    for (int heIdx = 0; heIdx < (int)mesh.halfEdges.size(); heIdx++) {
        HalfEdge& he = mesh.halfEdges[heIdx];
        int v1 = mesh.faces[he.face][heIdx % 3];
        int v2 = he.vertex;
        
        auto twinKey = edgeKey(v2, v1);
        auto it = edgeToHalfEdge.find(twinKey);
        
        // 需要找到反向的边
        for (int otherHeIdx = 0; otherHeIdx < (int)mesh.halfEdges.size(); otherHeIdx++) {
            if (otherHeIdx == heIdx) continue;
            HalfEdge& otherHe = mesh.halfEdges[otherHeIdx];
            
            int ov1 = mesh.faces[otherHe.face][otherHeIdx % 3];
            int ov2 = otherHe.vertex;
            
            if (ov1 == v2 && ov2 == v1) {
                he.twin = otherHeIdx;
                break;
            }
        }
    }
}

void BFFFlattener::identifyBoundaries() {
    for (int heIdx = 0; heIdx < (int)mesh.halfEdges.size(); heIdx++) {
        if (mesh.halfEdges[heIdx].twin == -1) {
            mesh.halfEdges[heIdx].isBoundary = true;
            
            int v1 = mesh.faces[mesh.halfEdges[heIdx].face][heIdx % 3];
            int v2 = mesh.halfEdges[heIdx].vertex;
            
            mesh.isBoundaryVertex[v1] = true;
            mesh.isBoundaryVertex[v2] = true;
        }
    }
}

bool BFFFlattener::flatten() {
    if (mesh.vertices.empty() || mesh.faces.empty()) {
        errorMsg = "Empty mesh";
        return false;
    }
    
    uvResult.clear();
    uvResult.resize(mesh.vertices.size() * 2, 0.0);
    
    // 对于简单情况，直接展开整个网格
    std::vector<int> allFaces;
    for (int i = 0; i < (int)mesh.faces.size(); i++) {
        allFaces.push_back(i);
    }
    
    std::vector<Vec2> uvs(mesh.vertices.size());
    std::map<int, int> identityMap;
    for (int i = 0; i < (int)mesh.vertices.size(); i++) {
        identityMap[i] = i;
    }
    
    if (!flattenPiece(allFaces, uvs, identityMap)) {
        return false;
    }
    
    // 复制结果
    for (int i = 0; i < (int)uvs.size(); i++) {
        uvResult[i * 2] = uvs[i].x;
        uvResult[i * 2 + 1] = uvs[i].y;
    }
    
    return true;
}

bool BFFFlattener::flattenPiece(const std::vector<int>& faceIndices,
                                 std::vector<Vec2>& uvs,
                                 std::map<int, int>& vertexMap) {
    if (faceIndices.empty()) return true;
    
    std::set<int> placedVertices;
    std::set<int> processedFaces;
    std::queue<int> faceQueue;
    
    // 从第一个面开始
    int firstFace = faceIndices[0];
    const auto& face = mesh.faces[firstFace];
    
    // 放置第一个三角形
    const Vec3& v0 = mesh.vertices[face[0]];
    const Vec3& v1 = mesh.vertices[face[1]];
    const Vec3& v2 = mesh.vertices[face[2]];
    
    double e01 = edgeLength(face[0], face[1]);
    double e02 = edgeLength(face[0], face[2]);
    double e12 = edgeLength(face[1], face[2]);
    
    // 第一个顶点在原点
    uvs[face[0]] = Vec2(0, 0);
    placedVertices.insert(face[0]);
    
    // 第二个顶点在x轴上
    uvs[face[1]] = Vec2(e01, 0);
    placedVertices.insert(face[1]);
    
    // 第三个顶点用余弦定理计算
    double cosA = (e01 * e01 + e02 * e02 - e12 * e12) / (2.0 * e01 * e02);
    cosA = std::max(-1.0, std::min(1.0, cosA));
    double sinA = std::sqrt(1.0 - cosA * cosA);
    uvs[face[2]] = Vec2(e02 * cosA, e02 * sinA);
    placedVertices.insert(face[2]);
    
    processedFaces.insert(firstFace);
    faceQueue.push(firstFace);
    
    // 构建面邻接关系
    std::map<std::pair<int, int>, std::vector<int>> edgeToFaces;
    for (int fIdx : faceIndices) {
        const auto& f = mesh.faces[fIdx];
        for (int i = 0; i < 3; i++) {
            int v1 = f[i];
            int v2 = f[(i + 1) % 3];
            edgeToFaces[edgeKey(v1, v2)].push_back(fIdx);
        }
    }
    
    // BFS展开
    while (!faceQueue.empty()) {
        int currentFace = faceQueue.front();
        faceQueue.pop();
        
        const auto& cf = mesh.faces[currentFace];
        
        // 检查相邻面
        for (int i = 0; i < 3; i++) {
            int ev1 = cf[i];
            int ev2 = cf[(i + 1) % 3];
            auto key = edgeKey(ev1, ev2);
            
            for (int neighborFace : edgeToFaces[key]) {
                if (processedFaces.count(neighborFace)) continue;
                
                const auto& nf = mesh.faces[neighborFace];
                
                // 找到共享边和新顶点
                int sharedV1 = -1, sharedV2 = -1, newV = -1;
                for (int v : nf) {
                    if (placedVertices.count(v)) {
                        if (sharedV1 == -1) sharedV1 = v;
                        else sharedV2 = v;
                    } else {
                        newV = v;
                    }
                }
                
                if (sharedV1 == -1 || sharedV2 == -1 || newV == -1) continue;
                
                // 计算新顶点位置
                const Vec2& p1 = uvs[sharedV1];
                const Vec2& p2 = uvs[sharedV2];
                
                double len12 = edgeLength(sharedV1, sharedV2);
                double len1n = edgeLength(sharedV1, newV);
                double len2n = edgeLength(sharedV2, newV);
                
                if (len12 < 1e-10) continue;
                
                // 使用余弦定理
                double cosAngle = (len12 * len12 + len1n * len1n - len2n * len2n) / (2.0 * len12 * len1n);
                cosAngle = std::max(-1.0, std::min(1.0, cosAngle));
                double sinAngle = std::sqrt(1.0 - cosAngle * cosAngle);
                
                // 计算方向
                Vec2 dir = (p2 - p1).normalize();
                Vec2 perp(-dir.y, dir.x);
                
                // 新顶点位置（选择使三角形朝向正确的方向）
                Vec2 newPos = p1 + dir * (len1n * cosAngle) + perp * (len1n * sinAngle);
                
                // 检查三角形朝向
                Vec2 e1 = p2 - p1;
                Vec2 e2 = newPos - p1;
                double cross = e1.x * e2.y - e1.y * e2.x;
                
                if (cross < 0) {
                    // 翻转
                    newPos = p1 + dir * (len1n * cosAngle) - perp * (len1n * sinAngle);
                }
                
                uvs[newV] = newPos;
                placedVertices.insert(newV);
                processedFaces.insert(neighborFace);
                faceQueue.push(neighborFace);
            }
        }
    }
    
    // 处理未放置的顶点
    for (int fIdx : faceIndices) {
        for (int v : mesh.faces[fIdx]) {
            if (!placedVertices.count(v)) {
                uvs[v] = Vec2(0, 0);
                placedVertices.insert(v);
            }
        }
    }
    
    // 共形优化
    optimizeConformal(uvs, faceIndices, vertexMap, 20);
    
    // 归一化UV坐标
    double minU = std::numeric_limits<double>::max();
    double maxU = std::numeric_limits<double>::lowest();
    double minV = std::numeric_limits<double>::max();
    double maxV = std::numeric_limits<double>::lowest();
    
    for (int fIdx : faceIndices) {
        for (int v : mesh.faces[fIdx]) {
            minU = std::min(minU, uvs[v].x);
            maxU = std::max(maxU, uvs[v].x);
            minV = std::min(minV, uvs[v].y);
            maxV = std::max(maxV, uvs[v].y);
        }
    }
    
    double scale = std::max(maxU - minU, maxV - minV);
    if (scale > 1e-10) {
        for (int fIdx : faceIndices) {
            for (int v : mesh.faces[fIdx]) {
                uvs[v].x = (uvs[v].x - minU) / scale;
                uvs[v].y = (uvs[v].y - minV) / scale;
            }
        }
    }
    
    return true;
}

void BFFFlattener::optimizeConformal(std::vector<Vec2>& uvs,
                                      const std::vector<int>& faceIndices,
                                      const std::map<int, int>& vertexMap,
                                      int iterations) {
    // 构建顶点邻接关系
    std::map<int, std::vector<int>> vertexNeighbors;
    for (int fIdx : faceIndices) {
        const auto& f = mesh.faces[fIdx];
        for (int i = 0; i < 3; i++) {
            int v = f[i];
            int vn = f[(i + 1) % 3];
            int vp = f[(i + 2) % 3];
            vertexNeighbors[v].push_back(vn);
            vertexNeighbors[v].push_back(vp);
        }
    }
    
    // 去重
    for (auto& pair : vertexNeighbors) {
        std::sort(pair.second.begin(), pair.second.end());
        pair.second.erase(std::unique(pair.second.begin(), pair.second.end()), pair.second.end());
    }
    
    // 迭代优化（拉普拉斯平滑 + 边长保持）
    double alpha = 0.5;
    
    for (int iter = 0; iter < iterations; iter++) {
        std::vector<Vec2> newUVs = uvs;
        
        for (auto& pair : vertexNeighbors) {
            int v = pair.first;
            const auto& neighbors = pair.second;
            
            if (neighbors.size() < 2) continue;
            
            // 计算邻居的平均位置
            Vec2 avg(0, 0);
            for (int n : neighbors) {
                avg = avg + uvs[n];
            }
            avg = avg * (1.0 / neighbors.size());
            
            // 拉普拉斯平滑
            newUVs[v] = uvs[v] * (1.0 - alpha) + avg * alpha;
        }
        
        uvs = newUVs;
    }
}

double BFFFlattener::computeAngle(const Vec3& a, const Vec3& b, const Vec3& c) {
    Vec3 ba = a - b;
    Vec3 bc = c - b;
    
    double dot = ba.dot(bc);
    double lenA = ba.length();
    double lenC = bc.length();
    
    if (lenA < 1e-10 || lenC < 1e-10) return 0;
    
    double cosAngle = dot / (lenA * lenC);
    cosAngle = std::max(-1.0, std::min(1.0, cosAngle));
    
    return std::acos(cosAngle);
}

double BFFFlattener::edgeLength(int v1, int v2) {
    return (mesh.vertices[v2] - mesh.vertices[v1]).length();
}

} // namespace bff

